libraries/eosiolib/contracts/eosio/key_value.hpp

```diff
@@ -462,6 +462,11 @@ namespace internal {
          return eosio::key_type{(char*)buffer, actual_value_size};
       }
 
+      bool valid() const {
+         eosio::check(itr_stat != status::iterator_erased, "Iterator was erased");
```

Comment:
This should just return true or false.  I wouldn't expect it to throw an assertion.  That ultimately means that you can't do something like `if(itr)` as it will either return true or assert, if that is the case then using `itr.valid()` and doing away with the `if` statement entirely would get you the same invariants.  Nonetheless, this is not what we want to occur, we want people to be able to query whether the iterator is valid without it throwing an error.

Also, this should be testing if the iterator is erased or it is and end iterator (there are 3 statuses for iterators).

Reply:
Updated